x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -315,10 +316,11 @@ private void onSearchResponse(SearchResponse searchResponse) {
             if (iterationResult.isDone()) {
                 logger.debug("Finished indexing for job [" + getJobId() + "], saving state and shutting down.");
 
-                // Change state first, then try to persist. This prevents in-progress
-                // STOPPING/ABORTING from
-                // being persisted as STARTED but then stop the job
-                doSaveState(finishAndSetState(), position.get(), this::onFinish);
+                // execute finishing tasks
+                onFinish(ActionListener.wrap(
+                        r -> doSaveState(finishAndSetState(), position.get(), () -> {}),
+                        e -> doSaveState(finishAndSetState(), position.get(), () -> onFailure(e))));
```

Comment:
since we already called onFinish I don't think we should call onFailure if saving the state throws an error. It's not ideal but I'd prefer that we ignore the error and just finish the job.